default/python/AI/freeorion_tools/_freeorion_tools.py

```diff
@@ -157,6 +157,26 @@ def wrapper(*args, **kwargs):
     return wrapper
 
 
+def cache_by_session_with_turnwise_update(function):
```

Comment:
`cache_by_turn` with arguments?

Reply:
No, a bit different, more just like I named it I think-- cache_by_turn keeps a history of the cache in foAI.foAIstate.  Since I didn't currently see a need for keeping that past history in the current use for this (especially for a potentially relatively large data structure) this cache only maintains a value for the current turn, and does not get stored in the savegame file.